sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,390 @@
+          ...
+        elif password:
+          ...
+        else
+          ...     
+```
+
+See Client [code](https://github.com/cyberark/conjur-api-python3/blob/main/conjur/api/client.py#L51) for reference.
+
+1. Multiple responsibilities and not easily expandable.
+
+   a. The constructor consumes 92 lines and handles fetching configuration and credential data, validating the input, configuring API calls, error handling, etc. This makes impacts code readability and should be broken up.
+
+   b. The constructor expects to have all configuration and credentials loaded by the time it is instantiated. If not, it attempts to fetch the missing information. This is problematic for flows such as `init`,  `login` that have not yet been run and therefore cannot use the Client.
+
+   c. In the CLI, we attempt to discover the organization's account for the user during the `init` flow. To get this information, we don't need to actually login to the Conjur server to get this information. We need to instead query `/info`. Because the Client is not expandable, this has forced us to bring over request logic to our logic code to avoid having to pass login information. See [InitLogic](https://github.com/cyberark/conjur-api-python3/blob/main/conjur/logic/login_logic.py#L54) for reference.
+
+2. Implict instead of explicit. 
+
+   a. The constuctor performs discovery depending on the parameter combination provided. A common example of this is when fetching credentials. The Client will search for credential details in the netrc only if a Keystore is not found. This is unnecessarily added work and as a result of this implicity, the user needs to be fully aware of the context of their environment.
+
+3. Areas of volatility are not encapsulated.
+
+   a. The constructor's parameters in the method signature are not encapsulated. It accepts 9 hanging parameters when they can instead be encapsulated into related objects. 
+
+   b. We can expect changes in the following areas in future iterations:
+
+   ​	i. Method of fetching configuration. Currently, we support fetching configuration details from the conjurrc file but this can change in the future and we will support methods in the future (via environment variables, etc).
+
+   ​	ii. Method of fetching credentials (via the netrc file or keystore). Currently, we support fetching credential data from the keystore or the netrc file.
+
+   Because of all these possible changes, we need to encapsulate them so that when changes arise, they are contained, without effecting the system.
+
+For these reasons, I propose we extract the configuration and credential information to their own objects to be created by the user before the Client is called. In those classes, all the collection and validations will be performed. As a result, the Client constructor will not be forced to be in charge of handle multiple responsibilities, the user will have to explicitly define the and the areas of volatility will be contained.
```

Comment:
What *is* the `Client` class responsibility?
It is not obvious, at least to me...

Reply:
Before or after my proposal?

Before, it happened 1. building config and cred objects 2. their validations 3. navigating requests according to the params it provided 5. and smart enough to know what to do when a param was missing

After my proposal, I plan for it to be an orchestrator, blind to the credential/configuration and command implementation details. It just gets inputs and shells out general instructions that will change according to the types of flow the user runs